modules/dxp/apps/segments/segments-experiment-web/src/main/resources/META-INF/resources/js/components/ClickGoalPicker/utils.es.js

```diff
@@ -39,15 +39,29 @@ export function stopImmediatePropagation(event) {
 /**
  * Returns all targetable elements within `element`
  *
- * Currently, that means all visible "a" and "button" elements which
- * have an "id".
+ * Currently, that means:
+ * - All visible "a" and "button" elements which have an "id".
+ * - An already selected target
  */
-export function getTargetableElements(element) {
-	const elements = element.querySelectorAll('a, button');
+export function getTargetableElements(element, selectedTarget) {
+
+	// Look for links or buttons
+
+	const elements = Array.from(element.querySelectorAll('a, button'));
+
+	// Other clickable element already selected
+
+	const target = document.getElementById(selectedTarget);
+
+	// Add selected target to elements
+
+	if (target && !elements.some((element) => target === element)) {
+		elements.push(target);
+	}
```

Comment:
So you have four linear operations here when two would suffice: `querySelectorAll()` involves a linear scan of the DOM, `Array.from` involves a linear loop (as you can see from looking at [a polyfill of it](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from#Polyfill)), then you're iterating a second time with `some()`, and finally a third time with `filter()`.

First a general comment: if you want set-like behavior (ie. "is this thing already a member of my set?" or "do I already have this thing") you should use a `Set` data structure. That means two of those loops (the `Array.from()` and the `some()`) would be better expressed using a `Set`:

```js
const elements = new Set(element.querySelectorAll('a, button'));

const target = document.getElementById(selectedTarget);

if (target) {
    elements.add(target);
}
```

(Note that `add()` on a `Set` won't add a duplicate of an already-present element because, by definition, that's the motivating behavior for using a `Set` in the first place.)

In this specific case, you have that fourth iteration (the `filter()`) to deal with as well, so we'd have to add basically the same code as we had prior to this PR:

```js
return Array.from(elements).filter((element) => {
    return element.id && _isVisible(element);
});
```

ie. down to 2 loops from 3.

We can't get rid of the `isVisible` check, but we could bake the `id` check into the `querySelectorAll`; ie: by using `element.querySelectorAll('a[id], button[id]')` instead of `element.querySelectorAll('a, button')`. And if you're going to do that, you could also obviate the whole need for the set by baking the `selectedTarget` look-up straight into the `querySelectorAll`.

Something like:

```js
const selector = ['a[id]', 'button[id]'];

if (selectedTarget) {
    selector.push(`#{selectedTarget}`);
}

const elements = document.querySelectorAll(selector.join());

return Array.from(elements).filter((element) => {
    return _isVisible(element);
});
```

(Disclaimer: untested, typed in browser etc... note I only put the `if (selectedTarget)` check in there because I am not sure whether it might be unset at this point...)

Reply:
Wow, what a nice explanation of what is going on and how to improve/simplify it. I really appreciate this. Thank you so much! I tried and everything works as expected.
When you create your AB Test with click goal, the click goal target is empty. If that check `if (selectedTarget)` is not there, `selector` will have a value like `#undefined`.